examples/ppi_unsup.py

```diff
@@ -0,0 +1,119 @@
+        super().__init__()
+        self.num_layers = num_layers
+        self.convs = torch.nn.ModuleList()
+        for i in range(num_layers):
+            in_channels = in_channels if i == 0 else hidden_channels
+            self.convs.append(SAGEConv(in_channels, hidden_channels))
+
+    def forward(self, x, edge_index):
+        for i, conv in enumerate(self.convs):
+            x = conv(x, edge_index)
+            if i != self.num_layers - 1:
+                x = x.relu()
+                x = F.dropout(x, p=0.5, training=self.training)
+        return x
+
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+model = GraphSAGE(train_dataset.num_features, 64, 2).to(device)
+optimizer = torch.optim.Adam(model.parameters(), lr=0.005)
+
+
+def train():
+    model.train()
+
+    total_loss = 0
+    for data in loader:
+        data = data.to(device)
+        optimizer.zero_grad()
+        out = model(data.x, data.edge_index)[:data.edge_label.shape[0]]
+        out, pos_out, neg_out = out.split(out.size(0) // 3, dim=0)
+
+        pos_loss = F.logsigmoid((out * pos_out).sum(-1)).mean()
+        neg_loss = F.logsigmoid(-(out * neg_out).sum(-1)).mean()
+        loss = -pos_loss - neg_loss
+        loss.backward()
+        optimizer.step()
+
+        total_loss += float(loss) * out.size(0)
+    return total_loss / len(loader.dataset)
+
+
+def get_representations_and_labels(model, graph_loader):
+    xs, ys = [], []
+    for data in graph_loader:
+        ys.append(data.y)
+        x, edge_index = data.x.to(device), data.edge_index.to(device)
+        out = model(x, edge_index)
+        xs.append(out)
+    return torch.cat(xs, dim=0).cpu(), torch.cat(ys, dim=0).cpu()
+
+
+@torch.no_grad()
+def test(train_loader, val_loader, test_loader):
+    model.eval()
+
+    # Create classifier
+    clf = MultiOutputClassifier(SGDClassifier(loss="log", penalty="l2"))
+
+    # Train classifier on train data
+    train_embeddings, train_labels = get_representations_and_labels(
+        model, train_loader)
+    clf.fit(train_embeddings, train_labels)
+    train_predictions = clf.predict(train_embeddings)
+    train_f1 = f1_score(train_labels, train_predictions, average='micro')
+
+    # Evaluate on validation set
+    val_embeddings, val_labels = get_representations_and_labels(
+        model, val_loader)
+    val_predictions = clf.predict(val_embeddings)
+    val_f1 = f1_score(val_labels, val_predictions, average='micro')
+
+    # Evaluate on testing set
+    test_embeddings, test_labels = get_representations_and_labels(
+        model, test_loader)
+    test_predictions = clf.predict(test_embeddings)
+    test_f1 = f1_score(test_labels, test_predictions, average='micro')
+
+    return train_f1, val_f1, test_f1
+
+
+for epoch in range(1, 51):
+    loss = train()
+    train_f1, val_f1, test_f1 = test(train_loader, val_loader, test_loader)
+    print('Epoch: {:03d}, Loss: {:.4f}, Train F1: {:.4f}, '
```

Comment:
`test()` might be expensive since we run a `clf` classifier on CPU in every epoch. Shall we only run this every 5th epoch?